fw/sock.c

```diff
@@ -812,12 +817,17 @@ ss_tcp_process_skb(struct sock *sk, struct sk_buff *skb, int *processed)
 		 * and the execution path should never reach here.
 		 */
 		BUG_ON(conn == NULL);
+
+		if (r) {
+			__kfree_skb(skb);
+			continue;
+		}
```

Comment:
Basically this is just `ss_skb_queue_purge()` with `ADJUST_PROCESSED_SKB()` inside and I'd prefer to keep the code similar to the original version with `ss_skb_queue_purge()` to separate the main logic from the error path and make the main logic more clear and both the code paths fast (w.o. the additional checks in both the cases).

Please add a comment to the commit message why do we need to account `tp->copied_seq` on the error path, in other words what happens otherwise with the socket (it seems this leads to the hung socket on shutdown).

One more thing: there are other places where `ss_skb_queue_purge()` is called, so don't we need to `ADJUST_PROCESSED_SKB()` inside the function to fit all the cases? If not, then probably we can just copy and paste code from ``ss_skb_queue_purge()` to `ss_tcp_process_skb()` with the comment why do we need to account `tp->copied_seq`.